lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -241,34 +157,88 @@ let next dis s =
     | [] -> Dis.stop dis s
     | r :: roots when not(Memory.contains s.base r) ->
       loop {s with roots}
-    | r :: roots when Span.mem s.visited r ->
+    | r :: roots when Visited.is_insn s.visited r ->
       loop {s with roots}
     | addr :: roots ->
-      let mem = match Span.upper_bound s.visited addr with
-        | None -> Memory.view ~from:addr s.base
-        | Some r1 -> Memory.range s.base addr r1  in
-      let mem =
-        Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
-      mem >>= fun mem -> Dis.jump dis mem {s with roots; addr} in
-  let visited = Span.add s.visited (s.addr, Dis.addr dis) in
-  loop {s with visited}
+      let mem = Memory.view ~from:addr s.base in
+      let mem = Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
+      mem >>= fun mem ->
+      Dis.jump dis mem {s with roots; addr} in
+  loop s
 
 let stop_on = [`Valid]
 
+let find_intersections s =
```

Comment:
This algorithm looks suboptimal and it computes more than returns, so it is a candidate for optimization. 
The algorithm is basically computing two groups of instructions:
1) those that share tails with any other instructions (represented with values that have more than one element), we need it because those instructions are barriers, and we need them in some efficient representation. This set is easily computed. 

2) a set of all instructions which intersect with some other instructions, but do not share tails (represented as values with singleton set). We need to in the later stage to ensure that all discovered instructions ended up in the CFG, and restart reconstruction on each. This set is hard to compute, however we can have it for free on stage 2, where we walk through the visited set and add instruction to CFG, at this step we should also remove the added instruction from the visited set. Therefore at the end we will have the sought  leftovers.